src/modules/Game.class.js

```diff
@@ -1,68 +1,344 @@
 'use strict';
 
-/**
- * This class represents the game.
- * Now it has a basic structure, that is needed for testing.
- * Feel free to add more props and methods if needed.
- */
+// const rows = document.querySelectorAll('tr');
+
 class Game {
-  /**
-   * Creates a new game instance.
-   *
-   * @param {number[][]} initialState
-   * The initial state of the board.
-   * @default
-   * [[0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0]]
-   *
-   * If passed, the board will be initialized with the provided
-   * initial state.
-   */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
+  constructor(
+    initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.board = initialState.map((row) => [...row]);
+    this.score = 0;
+    this.status = 'idle';
+    this.initialState = initialState;
+  }
+
+  getState() {
+    return this.board;
+  }
+
+  getScore() {
+    return this.score;
+  }
+
+  getStatus() {
+    return this.status;
+  }
+
+  start() {
+    this.status = 'playing';
+    this.newCell();
+    this.newCell();
+  }
+
+  restart() {
+    this.status = 'idle';
+    this.score = 0;
+
+    this.board = this.initialState.map((row) => [...row]);
+  }
+
+  newCell() {
+    const emptyCells = [];
+
+    for (let y = 0; y < 4; y++) {
+      for (let x = 0; x < 4; x++) {
+        if (!this.board[y][x]) {
+          emptyCells.push({ y: y, x: x });
+        }
+      }
+    }
+
+    if (emptyCells.length > 0) {
+      const val = Math.floor(Math.random() * emptyCells.length);
+
+      this.board[emptyCells[val].y][emptyCells[val].x] =
+        Math.random() < 0.9 ? 2 : 4;
+    }
+  }
+
+  checkGameState() {
+    if (this.board.find((row) => row.find((el) => el === 2048))) {
+      this.status = 'win';
+    } else if (this.checkIfLose()) {
+      this.status = 'lose';
+    }
+  }
+
+  checkIfLose() {
+    let lose = true;
+
+    if (this.board.find((row) => row.includes(0))) {
+      lose = false;
+    }
+
+    for (let y = 0; y < 4; y++) {
+      for (let x = 0; x < 4; x++) {
+        if (y < 3 && this.board[y][x] === this.board[y + 1][x]) {
+          lose = false;
+        }
+
+        if (x < 3 && this.board[y][x] === this.board[y][x + 1]) {
+          lose = false;
+        }
+      }
+    }
+
+    return lose;
```

Comment:
The method continues to check all cells even after determining that the game isn't lost.  Return false immediately upon finding an empty cell.